src/blindpsbt.cpp

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) 2017-2019 The Elements Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <blindpsbt.h>
+
+#include <hash.h>
+#include <primitives/transaction.h>
+#include <primitives/confidential.h>
+#include <psbt.h>
+#include <issuance.h>
+#include <random.h>
+#include <util/system.h>
+
+static secp256k1_context* secp256k1_blind_context = NULL;
+
+class Blind_PSBT_ECC_Init {
+public:
+    Blind_PSBT_ECC_Init() {
+        assert(secp256k1_blind_context == NULL);
+
+        secp256k1_context *ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+        assert(ctx != NULL);
+
+        secp256k1_blind_context = ctx;
+    }
+
+    ~Blind_PSBT_ECC_Init() {
+        secp256k1_context *ctx = secp256k1_blind_context;
+        secp256k1_blind_context = NULL;
+
+        if (ctx) {
+            secp256k1_context_destroy(ctx);
+        }
+    }
+};
+
+static Blind_PSBT_ECC_Init ecc_init_on_load;
+
+// Create surjection proof
+bool CreateAssetSurjectionProof(std::vector<unsigned char>& output_proof, const std::vector<secp256k1_fixed_asset_tag>& fixed_input_tags, const std::vector<secp256k1_generator>& ephemeral_input_tags, const std::vector<uint256>& input_asset_blinders, const uint256& output_asset_blinder, const secp256k1_generator& output_asset_tag, const CAsset& asset)
+{
+    int ret;
+    // 1 to 3 targets
+    size_t inputs_to_select = std::min((size_t)3, fixed_input_tags.size());
+    unsigned char randseed[32];
+    GetStrongRandBytes(randseed, 32);
+    size_t input_index;
+    secp256k1_surjectionproof proof;
+    secp256k1_fixed_asset_tag fixed_output_tag;
+    memcpy(&fixed_output_tag, asset.begin(), 32);
+    // Find correlation between asset tag and listed input tags
+    if (secp256k1_surjectionproof_initialize(secp256k1_blind_context, &proof, &input_index, &fixed_input_tags[0], fixed_input_tags.size(), inputs_to_select, &fixed_output_tag, 100, randseed) == 0) {
+        return false;
+    }
+    // Using the input chosen, build proof
+    ret = secp256k1_surjectionproof_generate(secp256k1_blind_context, &proof, &ephemeral_input_tags[0], ephemeral_input_tags.size(), &output_asset_tag, input_index, input_asset_blinders[input_index].begin(), output_asset_blinder.begin());
+    assert(ret == 1);
+    // Double-check answer
+    ret = secp256k1_surjectionproof_verify(secp256k1_blind_context, &proof, &ephemeral_input_tags[0], ephemeral_input_tags.size(), &output_asset_tag);
+    assert(ret != 0);
```

Comment:
For consistency, we should assert `== 1` like above

Reply:
Done here and elsewhere.